zebra/zebra_evpn_mh.c

```diff
@@ -1013,10 +1013,94 @@ static struct zebra_evpn_es_vtep *zebra_evpn_es_vtep_find(
 	return NULL;
 }
 
+static void zebra_evpn_es_df_change(struct zebra_evpn_es *es, bool new_non_df,
+				    const char *caller)
+{
+	bool old_non_df;
+
+	old_non_df = !!(es->flags & ZEBRA_EVPNES_NON_DF);
+
+	if (IS_ZEBRA_DEBUG_EVPN_MH_ES)
+		zlog_debug("df-change(%s) es %s old %s new %s", caller,
+			   es->esi_str, old_non_df ? "non-df" : "df",
+			   new_non_df ? "non-df" : "df");
+
+	if (old_non_df == new_non_df)
+		return;
+
+	if (new_non_df) {
+		es->flags |= ZEBRA_EVPNES_NON_DF;
+		/* XXX - Setup a dataplane DF filter to block BUM traffic */
+	} else {
+		es->flags &= ~ZEBRA_EVPNES_NON_DF;
+		/* XXX - clear the non-DF block filter */
+	}
+}
+
+static void zebra_evpn_es_run_df_election(struct zebra_evpn_es *es,
+					  const char *caller)
+{
+	struct listnode *node = NULL;
+	struct zebra_evpn_es_vtep *es_vtep;
+	bool new_non_df = false;
+
+	/* If the ES is not ready (i.e. not completely configured) there
+	 * is no need to setup the BUM block filter
+	 */
+	if (!(es->flags & ZEBRA_EVPNES_LOCAL)
+	    || !zmh_info->es_originator_ip.s_addr) {
+		zebra_evpn_es_df_change(es, new_non_df, caller);
+		return;
+	}
+
+	/* if oper-state is down DF filtering must be on. when the link comes
+	 * up again dataplane should block BUM till FRR has had the chance
+	 * to run DF election again
+	 */
+	if (!(es->flags & ZEBRA_EVPNES_OPER_UP)) {
+		new_non_df = true;
+		zebra_evpn_es_df_change(es, new_non_df, caller);
+		return;
+	}
+
+	for (ALL_LIST_ELEMENTS_RO(es->es_vtep_list, node, es_vtep)) {
+		/* Only VTEPs that have advertised the ESR can participate
+		 * in DF election
+		 */
+		if (!(es_vtep->flags & ZEBRA_EVPNES_VTEP_RXED_ESR))
+			continue;
+
+		/* If the DF alg is not the same we should fall back to
+		 * service-carving. But as service-carving is not supported
+		 * we will stop forwarding BUM
+		 */
+		if (es_vtep->df_alg != EVPN_MH_DF_ALG_PREF) {
```

Comment:
it would be good to have a log message in here to indicate to the user why the BUM traffic is not being forwarded.

Reply:
ack; I have added a warning log when a PE is added with a different df_alg -
+       if (df_alg != EVPN_MH_DF_ALG_PREF)
+               zlog_warn("remote es %s vtep %s add %s with unsupported df_alg %d",
+                          esi_to_str(esi, buf, sizeof(buf)),
+                          inet_ntoa(vtep_ip), esr_rxed ? "esr" : "", df_alg);
+